File: Python/112.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: TreeNode, targetSum: int) -> bool:
        return self.isTheSum(root, 0, targetSum)

    def isTheSum(self, root: TreeNode, currentSum: int, targetSum : int) -> bool:
        if not root: return False

        currentSum += root.val

        if not root.left and not root.right:
            if currentSum == targetSum:
                return True

        return self.isTheSum(root.left, currentSum, targetSum) or self.isTheSum(root.right, currentSum, targetSum)

# As noted, consider if the TreeNode has both positive and negative value at the sametime
# Which means we cannot know if we on the right path untill we reach each leaves.

File: CPP/112.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        return isTheSum(root, 0, targetSum);
    }

    bool isTheSum(TreeNode* root, int currentSum, int targetSum){
        // Not taking this path
        if (!root) return false;

        currentSum += root->val;

        if (!root->left && !root->right) {
             // if root is leaf
            if (currentSum == targetSum) return true;
        }
        return isTheSum(root->left, currentSum, targetSum)
             || isTheSum(root->right, currentSum, targetSum);
    }
};

// Consider the case that the elements in the TreeNode can be negative and positive at the same time.
// In that case, dont stop in the middle, but only consider the value when it is leaf.
